From f495d8428a8c8ddfcb3b85e3756e2582f46a6f23 Mon Sep 17 00:00:00 2001
From: Hieu Huynh <hieuh@os.amperecomputing.com>
Date: Fri, 12 May 2023 10:22:29 +0700
Subject: [PATCH] ampere: support BootProgress OemLastState

Support setting BootProgress OemLastState from the latest PostCodes entry.
This commit also supports to set LastState to Oem when dbus ProgressStages is Oem.

Gerrit: https://gerrit.openbmc.org/c/openbmc/bmcweb/+/58923

Upstream-Status: Pending

Signed-off-by: Hieu Huynh <hieuh@os.amperecomputing.com>
Change-Id: Ifbf32354d38bd68859b75b95040c6f09cd962cdc
---
 redfish-core/lib/systems.hpp | 56 ++++++++++++++++++++++++++++++++++++
 1 file changed, 56 insertions(+)

diff --git a/redfish-core/lib/systems.hpp b/redfish-core/lib/systems.hpp
index 5265436e..bcb32690 100644
--- a/redfish-core/lib/systems.hpp
+++ b/redfish-core/lib/systems.hpp
@@ -758,6 +758,12 @@ inline std::string dbusToRfBootProgress(const std::string& dbusBootProgress)
     {
         rfBpLastState = "OSRunning";
     }
+    else if (dbusBootProgress ==
+             "xyz.openbmc_project.State.Boot.Progress.ProgressStages."
+             "OEM")
+    {
+        rfBpLastState = "OEM";
+    }
     else
     {
         BMCWEB_LOG_DEBUG("Unsupported D-Bus BootProgress {}", dbusBootProgress);
@@ -853,6 +859,55 @@ inline void getBootProgress(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp)
         });
 }
 
+/**
+ * @brief Retrieves boot progress OemLastState of the system
+ *
+ * @param[in] aResp  Shared pointer for generating response message.
+ *
+ * @return None.
+ */
+inline void
+    getBootProgressOemLastState(const std::shared_ptr<bmcweb::AsyncResp>& aResp)
+{
+    crow::connections::systemBus->async_method_call(
+        [aResp](const boost::system::error_code ec,
+                const std::variant<std::tuple<std::vector<uint8_t>, std::vector<uint8_t>>>&
+                    code) {
+        if (ec)
+        {
+            return;
+        }
+
+        /* The OemLastState updated to the PostCodes Boot.Raw:
+         * https://gerrit.openbmc.org/c/openbmc/phosphor-dbus-interfaces/+/74633
+         * The type of the Boot.Raw value is struct[array[byte],array[byte]]
+         */
+        const std::tuple<std::vector<uint8_t>, std::vector<uint8_t>>*
+            codePtr =
+                std::get_if<std::tuple<std::vector<uint8_t>, std::vector<uint8_t>>>(&code);
+        if (codePtr == nullptr)
+        {
+            messages::internalError(aResp->res);
+            return;
+        }
+        std::string codeStr = bytesToHexString(std::get<0>(*codePtr));
+        /* Set "OemLastState": to "" if there is no boot progress code */
+        if (codeStr.empty())
+        {
+            codeStr = "";
+        }
+        else
+        {
+            codeStr = "0x" + codeStr;
+        }
+        aResp->res.jsonValue["BootProgress"]["OemLastState"] = codeStr;
+    },
+        "xyz.openbmc_project.State.Boot.Raw",
+        "/xyz/openbmc_project/state/boot/raw0",
+        "org.freedesktop.DBus.Properties", "Get",
+        "xyz.openbmc_project.State.Boot.Raw", "Value");
+}
+
 /**
  * @brief Retrieves boot progress Last Update of the system
  *
@@ -3135,6 +3190,7 @@ inline void
     getHostState(asyncResp);
     getBootProperties(asyncResp);
     getBootProgress(asyncResp);
+    getBootProgressOemLastState(asyncResp);
     getBootProgressLastStateTime(asyncResp);
     pcie_util::getPCIeDeviceList(asyncResp,
                                  nlohmann::json::json_pointer("/PCIeDevices"));
-- 
2.43.0

